src/workflows/recipes/components/PreferenceRolloutFields.js

```diff
@@ -0,0 +1,295 @@
+    } else if (typeof value === 'number') {
+      type = 'integer';
+    } else {
+      throw new Error(`Unrecognized preference type ${typeof value} (${value})`);
+    }
+    return {
+      name: pref.get('preferenceName'),
+      type,
+      value,
+      id: _nextId++,
+    };
+  });
+};
+
+export const deserializePreferenceRows = prefsList => {
+  return List(
+    // The reason for the filter is to remove any form argument items that are null
+    // or undefined which can happen when deleting a <FormItem> object.
+    prefsList.filter(row => !!row).map(row => {
+      return {
+        preferenceName: row.name,
+        value: row.value,
+      };
+    }),
+  );
+};
+
+@connectFormProps
+class PreferenceRolloutFields extends React.Component {
+  state = {
+    rows: serializePreferenceRows(this.props.recipeArguments.get('preferences')),
+  };
+
+  static propTypes = {
+    disabled: PropTypes.bool,
+    form: PropTypes.object.isRequired,
+    recipeArguments: PropTypes.instanceOf(Map),
+  };
+
+  static defaultProps = {
+    disabled: false,
+    recipeArguments: new Map(),
+  };
+
+  // Use a locally increasing number too to avoid the case were a new ID is one that
+  // it's already been.
+  _nextId = 0;
+
+  handleClickAddNewRow = () => {
+    this.setState(state => {
+      // Default to the same type as the last row.
+      // Useful if you add a bunch of booleans one after the other.
+      const type = state.rows.length ? state.rows[state.rows.length - 1].type : 'string';
+
+      // The nextId needs never be a number we've used before.
+      const ids = [...state.rows.values()].map(r => r.id);
+      const nextId = 1 + (ids.length ? Math.max(...ids) : 0) + this._nextId++;
+
+      return {
+        rows: state.rows.push({
+          name: '',
+          type: type,
+          value: DEFAULT_VALUES[type],
+          id: nextId,
+        }),
+      };
+    });
+  };
+
+  changeRow = row => {
+    this.setState(state =>
+      state.rows.map(r => {
+        if (r.id === row.id) {
+          return row;
+        }
+        return r;
+      }),
+    );
+  };
+
+  removeRow = row => {
+    // Don't allow this if there are currently only 1 rows.
+    if (this.state.rows.size <= 1) {
+      throw new Error('There must be at least one row left.');
+    }
+
+    this.setState(state => {
+      return {
+        rows: state.rows.filter(r => r.id !== row.id),
+      };
+    });
+  };
+
+  render() {
+    const { disabled, recipeArguments } = this.props;
+
+    return (
+      <div>
+        <Row>
+          <p className="action-info">Roll out a permanent preference change.</p>
+          <Col sm={24} md={11}>
+            <FormItem
+              label="Slug"
+              name="arguments.slug"
+              initialValue={recipeArguments.get('slug', '')}
+            >
+              <Input disabled={disabled} />
+            </FormItem>
+          </Col>
+        </Row>
+
+        <p className="action-info">Preferences</p>
```

Comment:
This doesn't look great. I can't see any other similar examples of headers for a group of inputs.

Reply:
yeah i'll do a quick style/css pass over this once it lands.